File: lib/spells/effects/Timed.cpp
/*
 * Timed.cpp, part of VCMI engine
 *
 * Authors: listed in file AUTHORS in main folder
 *
 * License: GNU General Public License v2.0 or later
 * Full text of license available in license.txt file, in main folder
 *
 */
#include "StdInc.h"

#include "Timed.h"
#include "../ISpellMechanics.h"

#include "../../battle/IBattleState.h"
#include "../../battle/CBattleInfoCallback.h"
#include "../../battle/Unit.h"
#include "../../bonuses/BonusParameters.h"
#include "../../json/JsonBonus.h"
#include "../../mapObjects/CGHeroInstance.h"
#include "../../networkPacks/PacksForClientBattle.h"
#include "../../networkPacks/SetStackEffect.h"
#include "../../serializer/JsonSerializeFormat.h"
#include "../../texts/CGeneralTextHandler.h"

VCMI_LIB_NAMESPACE_BEGIN

namespace spells
{
namespace effects
{

static void describeEffect(std::vector<MetaString> & log, const Mechanics * m, const std::vector<Bonus> & bonuses, const battle::Unit * target, const std::string & battleLogMessageSingular, const std::string & battleLogMessagePlural)
{
	for(const auto & bonus : bonuses)
	{
		if(bonus.type == BonusType::STACK_HEALTH && bonus.val < 0)
		{
			BonusList unitHealth = *target->getBonuses(Selector::type()(BonusType::STACK_HEALTH));

			auto oldHealth = unitHealth.totalValue();
			unitHealth.push_back(std::make_shared<Bonus>(bonus));
			auto newHealth = unitHealth.totalValue();

			//"The %s shrivel with age, and lose %d hit points."
			MetaString line;
			target->addText(line, EMetaText::GENERAL_TXT, 551);
			target->addNameReplacement(line);

			line.replaceNumber(oldHealth - newHealth);
			log.push_back(std::move(line));
			return;
		}
	}

	if(!battleLogMessagePlural.empty())
	{
		MetaString line;
		if(!battleLogMessageSingular.empty() && target->getCount() == 1)
			line.appendTextID(battleLogMessageSingular);
		else
			line.appendTextID(battleLogMessagePlural);
		target->addNameReplacement(line, boost::logic::indeterminate);
		log.push_back(std::move(line));
	}
}

void Timed::apply(ServerCallback * server, const Mechanics * m, const Target & target) const
{
	const bool describe = server->describeChanges();
	int32_t duration = m->getEffectDuration();

	std::vector<Bonus> converted;
	convertBonus(m, duration, converted);

	std::shared_ptr<const Bonus> peculiarBonus = nullptr;
	std::shared_ptr<const Bonus> addedValueBonus = nullptr;
	std::shared_ptr<const Bonus> fixedValueBonus = nullptr; 

	const auto *casterHero = dynamic_cast<const CGHeroInstance *>(m->caster);
	if(casterHero)
	{ 
		peculiarBonus = casterHero->getFirstBonus(Selector::typeSubtype(BonusType::SPECIAL_PECULIAR_ENCHANT, BonusSubtypeID(m->getSpellId())));
		addedValueBonus = casterHero->getFirstBonus(Selector::typeSubtype(BonusType::SPECIAL_ADD_VALUE_ENCHANT, BonusSubtypeID(m->getSpellId())));
		fixedValueBonus = casterHero->getFirstBonus(Selector::typeSubtype(BonusType::SPECIAL_FIXED_VALUE_ENCHANT, BonusSubtypeID(m->getSpellId())));
	}	
	//TODO: does hero specialty should affects his stack casting spells?

	SetStackEffect sse;
	BattleLogMessage blm;
	blm.battleID = m->battle()->getBattle()->getBattleID();
	sse.battleID = m->battle()->getBattle()->getBattleID();

	for(const auto & t : target)
	{
		std::vector<Bonus> buffer;
		std::copy(converted.begin(), converted.end(), std::back_inserter(buffer));

		const battle::Unit * affected = t.unitValue;
		if(!affected)
		{
			logGlobal->error("[Internal error] Invalid target for timed effect");
			continue;
		}

		if(!affected->alive())
			continue;

		if(describe)
			describeEffect(blm.lines, m, converted, affected, battleLogMessageSingular, battleLogMessagePlural);


		//Apply hero specials - peculiar enchants
		const auto tier = std::max(affected->creatureLevel(), 1); //don't divide by 0 for certain creatures (commanders, war machines)
		if(peculiarBonus && peculiarBonus->parameters)
		{
			si32 power = 0;
			switch (peculiarBonus->parameters->toNumber())
			{
			case 0: //normal
				switch (tier)
				{
				case 1:
				case 2:
					power = 3;
					break;
				case 3:
				case 4:
					power = 2;
					break;
				case 5:
				case 6:
					power = 1;
					break;
				}
				break;
			case 1: 
				//Coronius style specialty bonus.
				//Please note that actual Coronius isnt here, because Slayer is a spell that doesnt affect monster stats and is used only in calculateDmgRange
				break;
			}
			if(m->isNegativeSpell())
			{
				//negative spells like weakness are defined in json with negative numbers, so we need do same here
				power = -1 * power;
			}
			for(Bonus& b : buffer)
			{
				b.val += power;
			}

		}

		if(addedValueBonus && addedValueBonus->parameters)
		{
			for(Bonus & b : buffer)
			{
				b.val += addedValueBonus->parameters->toNumber();
			}
		}
		if(fixedValueBonus && fixedValueBonus->parameters)
		{
			for(Bonus & b : buffer)
			{
				b.val = fixedValueBonus->parameters->toNumber();
			}
		}

		if(cumulative)
			sse.toAdd.emplace_back(affected->unitId(), buffer);
		else
			sse.toUpdate.emplace_back(affected->unitId(), buffer);
	}

	if(!(sse.toAdd.empty() && sse.toUpdate.empty()))
		server->apply(sse);

	if(describe && !blm.lines.empty())
		server->apply(blm);
}

void Timed::convertBonus(const Mechanics * m, int32_t & duration, std::vector<Bonus> & converted) const
{
	int32_t maxDuration = 0;

	for(const auto & b : bonus)
	{
		Bonus nb(*b);

		//use configured duration if present
		if(nb.turnsRemain == 0)
			nb.turnsRemain = duration;
		vstd::amax(maxDuration, nb.turnsRemain);

		nb.sid = BonusSourceID(m->getSpellId()); //for all
		nb.source = BonusSource::SPELL_EFFECT;//for all

		//fix to original config: shield should display damage reduction
		if((nb.sid.as<SpellID>() == SpellID::SHIELD || nb.sid.as<SpellID>() == SpellID::AIR_SHIELD) && (nb.type == BonusType::GENERAL_DAMAGE_REDUCTION))
			nb.val = 100 - nb.val;
		//we need to know who cast Bind
		else if(nb.sid.as<SpellID>() == SpellID::BIND && nb.type == BonusType::BIND_EFFECT && m->caster->getHeroCaster() == nullptr)
			nb.parameters = std::make_shared<BonusParameters>(m->caster->getCasterUnitId());

		converted.push_back(nb);
	}

	//if all spell effects have special duration, use it later for special bonuses
	duration = maxDuration;
}

void Timed::serializeJsonUnitEffect(JsonSerializeFormat & handler)
{
	assert(!handler.saving);
	handler.serializeBool("cumulative", cumulative, false);
	{
		const JsonNode & messageNode = handler.getCurrent()["battleLogMessage"];
		if(messageNode.isStruct())
		{
			auto resolveField = [&](const std::string & field) -> std::string
			{
				const std::string & value = messageNode[field].String();
				if(value.empty())
					return {};
				if(value.at(0) == '@')
					return value.substr(1);
				return TextIdentifier("spell", spellScope, spellIdentifier, name, "battleLogMessage", field).get();
			};
			battleLogMessageSingular = resolveField("singular");
			battleLogMessagePlural = resolveField("plural");
		}
	}
	{
		auto guard = handler.enterStruct("bonus");
		const JsonNode & data = handler.getCurrent();

		for(const auto & p : data.Struct())
		{
			//TODO: support JsonSerializeFormat in Bonus
			auto guard = handler.enterStruct(p.first);
			const JsonNode & bonusNode = handler.getCurrent();
			auto b = JsonUtils::parseBonus(bonusNode);
			if (b)
				bonus.push_back(b);
			else
				logMod->error("Failed to parse bonus '%s'!", p.first);
		}
	}
}

}
}

VCMI_LIB_NAMESPACE_END

File: scripts/timed.lua
local Base = require("unitEffect")
local Script = setmetatable({}, {__index = Base})
Script.__index = Script

function Script:deepCopyBonus(b)
	local copy = {}
	for k, v in pairs(b) do
		copy[k] = v
	end
	return copy
end

function Script:convertBonuses(mechanics)
	local duration = mechanics:getEffectDuration()
	local spellKey = mechanics:getSpell():getJsonKey()
	local converted = {}

	for name, b in pairs(self.bonus or {}) do
		local nb = self:deepCopyBonus(b)

		if not nb.turns or nb.turns == 0 then
			nb.turns = duration
		end

		nb.sourceType = "SPELL_EFFECT"
		nb.sourceID = spellKey

		converted[name] = nb
	end

	return converted
end

function Script:applyHeroSpecialty(mechanics, buffer, unit)
	local hero = mechanics:getHeroCaster()
	if not hero then return end

	local spellKey = mechanics:getSpell():getJsonKey()
	local tier = math.max(unit:creatureLevel(), 1)

	local peculiar = hero:getBonuses(function(b)
		return b:getType() == "SPECIAL_PECULIAR_ENCHANT" and b:getSubtype() == spellKey
	end)
	if peculiar:size() > 0 then
		local mode = peculiar:getBonus(1):getParametersAsNumber()
		local power = 0
		if mode == 0 then
			if tier <= 2 then power = 3
			elseif tier <= 4 then power = 2
			elseif tier <= 6 then power = 1
			end
		end
		if mechanics:isNegative() then power = -power end
		if power ~= 0 then
			for _, nb in pairs(buffer) do
				nb.val = (nb.val or 0) + power
			end
		end
	end

	local addVal = hero:getBonuses(function(b)
		return b:getType() == "SPECIAL_ADD_VALUE_ENCHANT" and b:getSubtype() == spellKey
	end)
	if addVal:size() > 0 then
		local addAmount = addVal:getBonus(1):getParametersAsNumber()
		for _, nb in pairs(buffer) do
			nb.val = (nb.val or 0) + addAmount
		end
	end

	local fixedVal = hero:getBonuses(function(b)
		return b:getType() == "SPECIAL_FIXED_VALUE_ENCHANT" and b:getSubtype() == spellKey
	end)
	if fixedVal:size() > 0 then
		local fixedAmount = fixedVal:getBonus(1):getParametersAsNumber()
		for _, nb in pairs(buffer) do
			nb.val = fixedAmount
		end
	end
end

function Script:describeEffect(server, battle, unit, bonuses)
	-- Age spell: STACK_HEALTH bonus with negative val gets a custom message
	for _, nb in pairs(bonuses) do
		if nb.type == "STACK_HEALTH" and (nb.val or 0) < 0 then
			local healthBonuses = unit:getBonuses(function(b)
				return b:getType() == "STACK_HEALTH"
			end)
			local oldHealth = 0
			for i = 1, healthBonuses:size() do
				oldHealth = oldHealth + healthBonuses:getBonus(i):getVal()
			end
			local lost = oldHealth - (oldHealth + nb.val)
			local ageTextID = unit:getCount() == 1
				and "core.genrltxt.551"
				or  "core.genrltxt.552"
			server:appendLog(battle, {
				append         = { ageTextID },
				replaceStrings = { unit:getCreature():getNameTextID(unit:getCount()) },
				replaceNumbers = { lost }
			})
			return
		end
	end

	if not self.battleLogPlural or self.battleLogPlural == "" then return end

	local textID = (self.battleLogSingular and self.battleLogSingular ~= "" and unit:getCount() == 1) and self.battleLogSingular or self.battleLogPlural
	local nameTextID = unit:getCreature():getNameTextID(unit:getCount())
	server:appendLog(battle, {
		append         = { textID },
		replaceStrings = { nameTextID }
	})
end

function Script:apply(mechanics, server, target)
	local battle   = mechanics:getBattle()
	local describe = server:describeChanges()
	local converted = self:convertBonuses(mechanics)

	for _, dest in ipairs(target) do
		local unit = dest.unit
		if not unit or not unit:isAlive() then goto continue end

		local buffer = {}
		for name, nb in pairs(converted) do
			buffer[name] = self:deepCopyBonus(nb)
		end

		self:applyHeroSpecialty(mechanics, buffer, unit)

		if describe then
			self:describeEffect(server, battle, unit, buffer)
		end

		for _, nb in pairs(buffer) do
			server:addUnitBonus(battle, unit, nb, self.cumulative or false)
		end

		::continue::
	end
end

return Script
